### PR TITLE
AKCORE-49: Added acknowledgements Piggybacking in Share Fetch Request [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by chirag-wadhwa5**
This PR contains the implementation of Acknowledgements Piggybacking in Share Fetch Request. Required validations for acknowledgement batches in shareFetchRequest data has been added and the code has been thoroughly tested.
Reference: [AKCORE-49](https://confluentinc.atlassian.net/jira/software/c/projects/AKCORE/issues/AKCORE-49)

[AKCORE-49]: https://confluentinc.atlassian.net/browse/AKCORE-49?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 1]
```diff
@@ -1267,21 +1444,207 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+    val groupId = shareFetchRequest.data.groupId
+    val clientId = request.header.clientId
+    val memberId = shareFetchRequest.data().memberId()
+    val shareSessionEpoch = shareFetchRequest.data().shareSessionEpoch()
+
+    val shareFetchData = shareFetchRequest.shareFetchData(topicNames)
+    val forgottenTopics = shareFetchRequest.forgottenTopics(topicNames)
+
+    // TODO : replace this initialization when share fetch session metadata is sent by the client in the shareFetchRequest
+    val newReqMetadata : ShareFetchMetadata = new ShareFetchMetadata(Uuid.fromString(memberId), shareSessionEpoch)
```
Comment (adixitconfluent) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1488]: I noticed there are integration test failures with respect to Share Fetch. I debugged it locally and saw that the `memberId` getting populated inside `handleShareFetchRequest()` is coming out as empty string which could be causing the issue in this operation `Uuid.fromString(memberId)`

**Review by adixitconfluent — CHANGES_REQUESTED** [review]

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 2]
```diff
@@ -1072,31 +1073,226 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+            val topicIdPartition = new TopicIdPartition(
+              topic.topicId(),
+              new TopicPartition(topicNames.get(topic.topicId()), partition.partitionIndex())
+            )
+            var exceptionThrown = false
+            val acknowledgeBatches = new util.ArrayList[SharePartition.AcknowledgementBatch]()
+            breakable{
+              partition.acknowledgementBatches().forEach( batch => {
+                try {
+                  acknowledgeBatches.add(new SharePartition.AcknowledgementBatch(
```
Comment (adixitconfluent) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1179]: Why is the code calling `SharePartition` functions?  By the initial design, it should be calling functions in `SharePartitionManager`

**Review by adixitconfluent — CHANGES_REQUESTED** [review]

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 3]
```diff
@@ -1263,21 +1445,205 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+        }
+      })
+      isAcknowledgeDataPresent
+    }
+
+    var shareAcknowledgeResponse : ShareAcknowledgeResponse = null
+    var shareFetchResponse : ShareFetchResponse = null
+
+    // Handling the Acknowledgements from the ShareFetchRequest
+    val shareSessionError: Errors = sharePartitionManager.acknowledgeShareSessionCacheUpdate(groupId, Uuid.fromString(memberId), shareSessionEpoch)
```
Comment (AndrewJSchofield) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1511]: This piece seems wrong. When sending the first request in a session, you would expect no acknowledge data in the the fetch request, but you would also expect the share session would not be in the cache yet. So, we get an error code for the acknowledgements in this case, when I would say it should be Errors.NONE.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
The error handling for piggybacked acknowledgements in ShareFetch isn't quite right I think.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 4]
```diff
@@ -1263,21 +1445,205 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+    val newReqMetadata : ShareFetchMetadata = new ShareFetchMetadata(Uuid.fromString(memberId), shareSessionEpoch)
+    val shareFetchContext = sharePartitionManager.newContext(groupId, shareFetchData, forgottenTopics, topicNames, newReqMetadata)
+
+    // Handling the Fetch from the ShareFetchRequest
+    // Fetching should not proceed under any one of the following scenarios -
+    // 1. If shareAcknowledgeResponse has Errors.INVALID_REQUEST error code
+    // 2. If this is a Final Fetch request (shareSessionEpoch == ShareFetchMetadata.FINAL_EPOCH) and shareFetchData is not empty
+    // 3. If this is a Final Fetch request (shareSessionEpoch == ShareFetchMetadata.FINAL_EPOCH) and forgottenTopics is not empty
+    if (shareAcknowledgeResponse.data().errorCode() ==  Errors.INVALID_REQUEST.code()
+      || (shareSessionEpoch == ShareFetchMetadata.FINAL_EPOCH && (!shareFetchData.isEmpty || !forgottenTopics.isEmpty))) {
```
Comment (AndrewJSchofield) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1543]: In the case where it is FINAL_EPOCH, we do not want to fetch at all. We might have acknowledgements to process. It doesn't matter whether shareFetchData or forgottenTopics is empty. We process the acknowledgements, close the share session and do not fetch data.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 5]
```diff
@@ -1263,21 +1445,205 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+      shareFetchResponse = shareFetchRequest.getErrorResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, Errors.INVALID_REQUEST.exception) match {
+        case response: ShareFetchResponse => response
+        case _ => null
+      }
+    } else{
```
Comment (AndrewJSchofield) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1548]: Tiny formatting nit.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 6]
```diff
@@ -4235,68 +4601,68 @@ class KafkaApis(val requestChannel: RequestChannel,
...
-          nodeEndpoints.values.toList.asJava
-        )
-      }
-      unconvertedShareAcknowledgeResponse = new ShareAcknowledgeResponse(
-        ShareAcknowledgeResponse.toMessage(Errors.NONE, 0, partitions.entrySet().iterator(), Collections.emptyList())
-      )
-
-      requestHelper.sendMaybeThrottle(request, createResponse())
-    }
+//    def processResponseCallback(
```
Comment (AndrewJSchofield) [inline core/src/main/scala/kafka/server/KafkaApis.scala:4604]: This large comment ought to be removed before we merge the code.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 7]
```diff
@@ -1072,31 +1073,226 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+    val groupId = shareFetchRequest.data().groupId()
+    val clientId = request.header.clientId
+    val memberId = shareFetchRequest.data().memberId()
 
-    if (!config.isShareGroupEnabled) {
-      // The API is not supported when the configuration `group.share.enable` has not been set
-      requestHelper.sendMaybeThrottle(request, shareFetchRequest.getErrorResponse(Errors.UNSUPPORTED_VERSION.exception))
-      CompletableFuture.completedFuture[Unit](())
-      return
+    if (!authHelper.authorize(request.context, READ, GROUP, groupId)) {
```
Comment (AndrewJSchofield) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1148]: I would have expected all of the auth checking to be at the start of `handleShareFetchRequest`.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 1 above)
Comment (AndrewJSchofield) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: This has been fixed in the client I think. Rebasing on the kip-932 branch should resolve it.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 2 above)
Comment (adixitconfluent) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1179]: nevermind I see that you are only utilizing a data structure but not the functionality, so it should be fine.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 3 above)
Comment (chirag-wadhwa5) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1511]: Thanks for the review, I will make this change in the next commit

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 4 above)
Comment (chirag-wadhwa5) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1543]: [This](https://confluentinc.atlassian.net/wiki/spaces/AK/pages/3334014224/KIP-932+Share+sessions) doc says that if session epoch is -1 (final fetch request), and if it contains topics to add or forget, the request should fail with the error INVALID_REQUEST

File: core/src/main/scala/kafka/server/BrokerServer.scala [hunk 8]
```diff
@@ -394,7 +394,8 @@ class BrokerServer(
 
       // TODO : add the config max.incremental.share.fetch.session.cache.slots and the variable MIN_INCREMENTAL_SHARE_FETCH_SESSION_EVICTION_MS
       //  which should be used to initialize shareFetchSessionCache
-      val shareFetchSessionCache : ShareSessionCache = new ShareSessionCache(0, 0)
+      val shareFetchSessionCache : ShareSessionCache = new ShareSessionCache(config.maxIncrementalFetchSessionCacheSlots,
```
Comment (AndrewJSchofield) [inline core/src/main/scala/kafka/server/BrokerServer.scala:397]: The comment above should be removed. The KIP says the maximum size is `group.share.max.groups * group.share.max.size`. The eviction timeout has no specific configuration for now. The KIP says 120,000 milliseconds (2 minutes).

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 9]
```diff
@@ -1284,6 +1274,7 @@ class KafkaApis(val requestChannel: RequestChannel,
                                        topicNames : util.Map[Uuid, String],
                                        sharePartitionManager : SharePartitionManager,
                                        shareFetchContext : ShareFetchContext,
+                                       unauthorizedTopicPartitions: mutable.Set[TopicIdPartition]
                                       ): ShareFetchResponse = {
 
     val shareFetchRequest = request.body[ShareFetchRequest]
```
Comment (apoorvmittal10) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1280]: Shouldn't we have common method to handle acknowledgement from ShareFetch and ShareAcknowledge as most of the functionality would remain same?

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 10]
```diff
@@ -1502,43 +1492,62 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+    val authorizedTopics = authHelper.filterByAuthorized(
+      request.context,
+      READ,
+      TOPIC,
+      shareFetchData.asScala
+    )(_._1.topicPartition.topic)
+    shareFetchData.forEach((tp, _) => {
+      if (!authorizedTopics.contains(tp.topic))
+        unauthorizedTopicPartitions.add(tp)
+    })
```
Comment (apoorvmittal10) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1505]: Sorry I might didn't understand, we already have list of unauthorized topics why filtering again?

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 11]
```diff
@@ -1502,43 +1492,62 @@ class KafkaApis(val requestChannel: RequestChannel,
...
-        shareAcknowledgeResponse = shareFetchRequest.getErrorAcknowledgeResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, Errors.INVALID_REQUEST.exception) match {
-          case response: ShareAcknowledgeResponse => response
-          case _ => null
+    if (!authHelper.authorize(request.context, READ, GROUP, groupId)) {
+      shareAcknowledgeResponse = shareFetchRequest.getErrorAcknowledgeResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, Errors.GROUP_AUTHORIZATION_FAILED.exception) match {
+        case response: ShareAcknowledgeResponse => response
+        case _ => null
+      }
+    } else {
+      val isAcknowledgeDataPresent = isAcknowledgeDataPresentInFetchRequest()
```
Comment (apoorvmittal10) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1517]: Sorry but I don't see this method (`isAcknowledgeDataPresentInFetchRequest`) defined, neither in existing KafkaApis code. What I am missing? Also shouldn't this be first check prior proceeding with any acknowledgement code?

File: core/src/main/scala/kafka/server/BrokerServer.scala (hunk 8 above)
Comment (chirag-wadhwa5) [reply core/src/main/scala/kafka/server/BrokerServer.scala]: Thanks for pointing it our, made this change in the latest commit

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 4 above)
Comment (AndrewJSchofield) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: By "topics to add", I mean if there are topics in the request which do not exist in the session, then fail it. But you can legitimately have entries in the topics array that are just acknowledgements for topic-partitions which are already in the session. Those should be processed, and then the session should be removed from the cache.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 10 above)
Comment (chirag-wadhwa5) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: That was really stupid, sorry about that. Made the change in. the latest commit

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 11 above)
Comment (chirag-wadhwa5) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1512]: I defined it inside the handleShareFetch only.
We don't need to this check at all if for example the authorisation fails for reading the group.

**Review by AndrewJSchofield — APPROVED** [review]
lgtm. I expect we'll need some adjustments as we test more, but it's a big step forward.

**Review by AndrewJSchofield — APPROVED** [review]
Thanks for the PR. lgtm

**Pull request merged**